### ISSUE TITLE
doc string for __init__

### ISSUE THREAD

**Issue by ericvsmith** [ericvsmith/dataclasses#32]
From a discussion with @raymondh:

Should we auto-generate some sort of doc string for `__init__`, or allow the user to specify one (maybe as a param to `@dataclass`)?

I'm not sure a generated one wouldn't have too much noise to be useful.

**Comment by ilevkivskyi** [comment]
> Should we auto-generate some sort of doc string

I think yes. This is the case for named tuples, and it helps for libraries that use named tuples. I think the behavior should be: auto-generate a docstring if ``doc=None`` keyword is not given, otherwise just use the ``doc``. Btw, I think we could provide the possibility to add docstrings for fields:
```python
@dataclass
class Char:
    x: int
    y: int
    label: Optional[str] = field(default=None, doc='Shows name and health points')
```
(These will also help with autogeneratig docstring for ``__init__``.)

**Comment by taleinat** [comment]
Do we really need to? `help()` knows to generate a generic description of `__init__` if it has no doc-string (see below). If `__init__` doesn't do anything other than set parameters as attributes, IMO it doesn't need a doc-string, and if it does then it wouldn't be the auto-generated one.

The exception is when a post-init method is used; in that case we could just copy the post-init method's doc-string if it is set.

The further suggestion of "building" a doc-string by aggregating per-field docs is another matter.

```python
>>> class A:
...     def __init__(zelf, x, y, z):
...         pass
...
>>> A.__init__.__doc__
>>> help(A.__init__)
Help on function __init__ in module __main__:

__init__(zelf, x, y, z)
    Initialize self.  See help(type(self)) for accurate signature.

>>> A.__doc__
>>> help(A)
Help on class A in module __main__:

class A(builtins.object)
 |  A(x, y, z)
 |
 |  Methods defined here:
 |
 |  __init__(zelf, x, y, z)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |
 |  ----------------------------------------------------------------------
 |  Data descriptors defined here:
 |
 |  __dict__
 |      dictionary for instance variables (if defined)
 |
 |  __weakref__
 |      list of weak references to the object (if defined)

```

**Comment by taleinat** [comment]
Example using current version of `dataclasses`:
```python
@dataclass
class B:
    x:str = "1"
    y:int = field(default=2)
    z:int = 3
    
help(B.__init__)
Help on function __init__:
__init__(self, x:str='1', y:int=2, z:int=3) -> None
    B.__init__(self, x, y, z)
```

**Comment by ericvsmith** [comment]
Thanks for looking at this, @taleinat.

I agree that the current behavior is pretty good. The only quibble I have is with `help(B)`, vs. `help(B.__init__)`. I'm unlikely to ever type the latter.

Contrast with `namedtuple`, which tells you right away how to call it:
```
>>> C = namedtuple('C', 'x y z')
>>> help(C)
Help on class C in module __main__:

class C(builtins.tuple)
 |  C(x, y, z)
 |
...
```
Versus with a dataclass:
```
>>>help(B)

Help on class B in module __main__:

class B(builtins.object)
 |  Methods defined here:
 |
 |  __eq__(self, other)
 |
 |  __ge__(self, other)
 |
 |  __gt__(self, other)
 |
 |  __init__(self, x:str='1', y:int=2, z:int=3) -> None
 |
...
```
Note that the "how to call this" isn't first, and instead of `B(...)`, it's several functions down, under `__init__(self, ...)`. In an ideal world, it would look more like namedtuple.

I don't know how easy it would be to leverage some of the auto-generated help output. Maybe it's not worth the hassle, or maybe it's a quality of implementation issue we could deal with later.

I would also tend toward punting the per-field docstrings down the field, and tackle it at a later date.

**Comment by ericvsmith** [comment]
@taleinat: I think it's worth doing if there's some place we can call to get the correct string. I don't think it's worth doing if we have to reinvent the wheel and build up the signature ourselves.

**Comment by taleinat** [comment]
I'm taking a look at the `pydoc` module, which is what `help()` uses.

**Comment by taleinat** [comment]
See initial implementation based on the same utility used internally by `help()`.

**Comment by ilevkivskyi** [comment]
I haven't followed this closely, but why is the docstring for ``__init__`` not follow the PEP-8 rules? I mean it should be
```
__init__(self, x: str = '1', y: int = 2, z: int = 3) -> None
```
not
```
__init__(self, x:str='1', y:int=2, z:int=3) -> None
```
(note spaces)

**Comment by taleinat** [comment]
@ilevkivskyi, PEP 8 specifically makes an exception for keyword arguments and default values. It's a few bullets down under "Other Recommendations":

> Don't use spaces around the = sign when used to indicate a keyword argument or a default parameter value.
>
> Yes:
>
> ```def complex(real, imag=0.0):
    return magic(r=real, i=imag)```
>
> No:
>
> ```def complex(real, imag = 0.0):
    return magic(r = real, i = imag)```

**Comment by ilevkivskyi** [comment]
@taleinat 
If you would scroll a bit more you could read this:
```
When combining an argument annotation with a default value, use spaces around the = sign
(but only for those arguments that have both an annotation and a default).

Yes:

def munge(sep: AnyStr = None): ...
def munge(input: AnyStr, sep: AnyStr = None, limit=1000): ...

No:

def munge(input: AnyStr=None): ...
def munge(input: AnyStr, limit = 1000): ...
```

**Comment by taleinat** [comment]
@ilevkivskyi Wow, I hadn't seen that part! I guess it was added when annotations were.

This code just uses `str(inspect.signature(C))`, so perhaps that should be made to generate PEP-8-compliant signature strings?

**Comment by ericvsmith** [comment]
I think that this isn't something we should care about here.

It's what inspect.signature() and help() display:

```
>>> def __init__(self, x: str = '1', y: int = 2, z: int = 3) -> None: pass
... 
>>> import inspect
>>> str(inspect.signature(__init__))
"(self, x:str='1', y:int=2, z:int=3) -> None"
>>> help(__init__)
Help on function __init__ in module __main__:

__init__(self, x:str='1', y:int=2, z:int=3) -> None

>>> 
```

**Comment by ilevkivskyi** [comment]
I will open an issue at b.p.o. about this.

**Comment by ilevkivskyi** [comment]
https://bugs.python.org/issue32018

**Comment by taleinat** [comment]
So @ericvsmith, any more comments on the PR now that it has been updated, or can it go in?

**Comment by ericvsmith** [comment]
Closed via #64.

**Comment by ericvsmith** [comment]
Thanks @taleinat!
